app/models/docker_registry.rb

```diff
@@ -20,33 +20,36 @@ def apply(settings_params)
     # create or update DockerRegistry model
     def configure_registry(registry)
       errors      = []
-      url         = registry["docker_registry_url"]
-      cert        = registry["docker_registry_certificate"]
-      mirror      = registry["docker_registry_mirror"]
+      url         = registry["url"]
+      cert        = registry["certificate"]
+      mirror      = registry["mirror"]
 
-      if cert.present?
-        certificate = Certificate.find_or_create_by certificate: cert.strip
-        errors << "Failed to validate certificate" unless certificate.persisted?
-      end
-
-      service = DockerRegistry.find_or_create_by(url: url) do |r|
+      registry = DockerRegistry.find_or_create_by(url: url) do |r|
         r.mirror = mirror
-        r.certificate = certificate
       end
 
-      unless service.persisted?
+      unless registry.persisted?
         errors << "Registry url #{url} doesn't match a docker registry pattern"
       end
 
+      if cert.present?
+        certificate = Certificate.find_or_create_by(certificate: cert.strip)
+        errors << "Failed to validate certificate" unless certificate.persisted?
+
+        CertificateService.create(service: registry, certificate: certificate) if certificate.persisted?
+      elsif registry.certificate.present?
+        registry.certificate_service.destroy
+      end
```

Comment:
Also, as an improvement, we could try to refactor this a bit to use a transaction and rollback everything if any of the steps breaks, that way we ensure our data is consistent.